src/transport/SecureMessageCodec.cpp

```diff
@@ -43,22 +43,23 @@ CHIP_ERROR Encrypt(const CryptoContext & context, CryptoContext::ConstNonceView
     VerifyOrReturnError(!msgBuf->HasChainedBuffer(), CHIP_ERROR_INVALID_MESSAGE_LENGTH);
     VerifyOrReturnError(msgBuf->TotalLength() <= kMaxAppMessageLen, CHIP_ERROR_MESSAGE_TOO_LONG);
 
-    static_assert(std::is_same<decltype(msgBuf->TotalLength()), uint16_t>::value,
-                  "Addition to generate payloadLength might overflow");
-
     ReturnErrorOnFailure(payloadHeader.EncodeBeforeData(msgBuf));
 
-    uint8_t * data    = msgBuf->Start();
-    uint16_t totalLen = msgBuf->TotalLength();
+    uint8_t * data  = msgBuf->Start();
+    size_t totalLen = msgBuf->TotalLength();
 
     MessageAuthenticationCode mac;
     ReturnErrorOnFailure(context.Encrypt(data, totalLen, data, nonce, packetHeader, mac));
 
     uint16_t taglen = 0;
     ReturnErrorOnFailure(mac.Encode(packetHeader, &data[totalLen], msgBuf->AvailableDataLength(), &taglen));
 
-    VerifyOrReturnError(CanCastTo<uint16_t>(totalLen + taglen), CHIP_ERROR_INTERNAL);
-    msgBuf->SetDataLength(static_cast<uint16_t>(totalLen + taglen));
+#if CHIP_SYSTEM_CONFIG_USE_LWIP
+    VerifyOrReturnError(totalLen + taglen <= UINT16_MAX, CHIP_ERROR_INTERNAL);
+#else  // CHIP_SYSTEM_CONFIG_USE_LWIP
+    VerifyOrReturnError(totalLen + taglen <= UINT32_MAX, CHIP_ERROR_INTERNAL);
+#endif // CHIP_SYSTEM_CONFIG_USE_LWIP
```

Comment:
Why are we ending up with ifdefs here?  What situation are we actually trying to guard against, and why does it need an ifdef?

SetDataLength() takes a size_t, yes?  Why can't we just do the addition and pass in our size_t?  The packet buffer will clamp the incoming length itself as needed, so if we want to be 100% sure that the set succeeded, we could just check that the data length is what we expect after the set call.

Again, the old code was trying to handle two things:

1. Overflow on the `totalLen + taglen` addition.  The new code just assumes that will never happen, and maybe that's OK, given that `taglen` is small and `totalLen` fits in sane-sized memory?  But if we ever start having 4GB packet buffers we need to worry about this.
2. Making sure the cast to uint16_t did not drop data.  That's not an issue here, because we are no longer casting to a smaller type.   The only place data might get dropped now is if SetDataLength() clamps, which was an unhandled case in the old code....

Reply:
I guess the checks here are pre-emptive sanity checks for the length. The reason for the ifdef is because LwIP uses u16 throughout. For Linux sockets, UINT32_MAX is a reasonable loose upper bound. I could remove these but it seemed better to have the bound check apriori rather than comparing against the size after SetDataLength(), no? I noticed that SetDataLength() internally has the UINT16_MAX check for LwIP. Maybe, I can add a UINT32_MAX check for Linux sockets there and drop these checks?

Reply:
Removed the checks from here.